rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,296 @@
+# RFC 120 object Service
+* Comments: [#116](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/120)
+* Proposed by: @jhford
+
+# Summary
+
+Taskcluster tasks often produce artifacts during the course of being run.
+Examples include browser distribution archives, json metadata and log files.
+We currently store artifacts in a single region of S3 and split management of
+artifacts between the Queue and Cloud Mirror.  This project is to design an
+object storage api which drastically simplifies the management of objects from
+the Queue's perspective while at the same time handling the cross-region
+replication and cache management of Cloud Mirror as well as enabling
+cross-cloud support.
+
+The object service should not be directly exposed to end users as a Rest API.
+It will instead be a generalised API that the Queue uses to facilitate artifact
+creation from tasks.  In other words, the Queue gates operations to the object
+service.  The service will be exposed to endusers by means of a `Location`
+header value on a `300` series redirect or by relaying responses.
+
+The Queue will now store only the name of objects and would treat the name of
+an object as an opaque identifier and defer all management of the actual bytes
+to this system.  The Queue would also stop doing IP to object store mappings.
+
+This RFC is to decide on what the objects service is and its interface, not
+implementation details.
+
+## Motivation
+
+* We want the Queue to be easier to understand, develop and maintain
+* We want objects to be easier to understand, develop and maintain
+* We want better object handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of objects into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* Auth will not be done with scopes and tc-auth, but a shared secret between
+  frontend service and objects, with JWT being a possibility
```

Comment:
I think it would be useful to describe the sequence of HTTP requests between components first, so it is clearer which HTTP requests require authorization, and how authentication is achieved.

Reply:
The rough design so far is that all gets, deletes and creates would need to be authenticated.  At that point, it's probably best to force all requests to use the same authentication and authorization.